content/python/hardware_components.py.md

```diff
@@ -0,0 +1,172 @@
+        table.add_row(['Location', result['datacenter']['longName']])
+        table.add_row(['Serial #', result['serialNumber']])
+        table.add_row(['Billing', "Hourly" if result['billingItem']['hourlyFlag'] else "Monthly"])
+        table.add_row(['Last transaction', result['lastTransaction']['transactionGroup']['name']])
+        table.add_row(['cores', result['processorPhysicalCoreAmount']])
+        table.add_row(['memory', result['memoryCapacity']])
+        table.add_row(['public_ip', result['primaryIpAddress']])
+        table.add_row(['private_ip', result['primaryBackendIpAddress']])
+        table.add_row(['os', result['operatingSystem']['softwareLicense']['softwareDescription']['manufacturer']])
+        table.add_row(['os_version', result['operatingSystem']['softwareLicense']['softwareDescription']['version']])
+        print(table)
+
+    def list_hardware_components(self, server_id):
+        mask = 'mask[activeComponents[hardwareComponentModel[hardwareGenericComponentModel[hardwareComponentType]]]]'
+        result = self.client['SoftLayer_Hardware_Server'].getObject(id=server_id, mask=mask)
+
+        table = PrettyTable()
+        table.title = "Hardware Components"
+        table.field_names = ['id', 'capacity', 'manufacturer', 'name', 'version', 'hardwareComponentType']
+
+        capacity = None
+        component_type = None
+        for component in result['activeComponents']:
+            if 'hardwareGenericComponentModel' in component['hardwareComponentModel']:
+                capacity = component['hardwareComponentModel']['hardwareGenericComponentModel']['capacity']
+                component_type = component['hardwareComponentModel']['hardwareGenericComponentModel'][
+                    'hardwareComponentType']['type']
+                table.add_row([component['hardwareComponentModel']['id'],
+                               capacity,
+                               component['hardwareComponentModel']['manufacturer'],
+                               component['hardwareComponentModel']['name'],
+                               component['hardwareComponentModel']['version'],
+                               component_type
+                               ])
+
+        print(table)
+
+    def list_hardware_component_firmware(self, server_id, hardware_component_id):
+        mask = 'filteredMask[activeComponents[hardwareComponentModel[firmwares]]]'
+        object_filter = {
+            "hardware": {
+                "id": {
+                    "operation": server_id
+                },
+                "activeComponents": {
+                    "hardwareComponentModel": {
+                        "id": {
+                            "operation": hardware_component_id
+                        }
+                    }
+                }
+            }
+        }
+
+        result = self.client['SoftLayer_Account'].getHardware(mask=mask, filter=object_filter)
```

Comment:
If you know the server Id, I'd prefer this example use `SoftLayer_Hardware_Server.getObject` directly as I think that sets a better example for users.

Reply:
Done.